### PR TITLE
Feature/alive 

### PR DESCRIPTION
**Pull request by hoonkii**
resolved: #[issue number]

details:
Implement member_map/ Alive and Delete AddMember
Implement Alive TestCase
Modify Test Case

- [X] Test case

### CODE REVIEW

File: member_map.go [hunk 1]
```diff
@@ -195,9 +183,42 @@ func override(newMem Member, existingMem Member) Member {
 }
 
 // Update member and update waitingMembers
-// todo
-func (m *MemberMap) UpdateMember(member Member) error {
-	return nil
+// if Member status in MemberList is changed return true else return false
```
Comment (junbeomlee) [inline member_map.go]: Change AliveObject to AliveMsg

File: member_map.go [hunk 2]
```diff
@@ -73,6 +71,19 @@ type Member struct {
 	// got suspicion message, that member can increments incarnation
 	Incarnation uint32
 }
+type AliveMessageID struct {
+	ID string
+}
+
+type AliveMessage struct {
+	ID AliveMessageID
+
```
Comment (junbeomlee) [inline member_map.go:76]: Change `ID AliveMessageID` to `ID string`

File: member_map.go (hunk 1 above)
Comment (hoonkii) [reply member_map.go]: Ok

File: member_map.go (hunk 2 above)
Comment (hoonkii) [reply member_map.go:76]: OK

File: member_map.go [hunk 3]
```diff
@@ -198,10 +190,48 @@ func override(newMem Member, existingMem Member) Member {
...
+	// if Member is not exist in MemberList
+	if !ok {
+		m.members[MemberID{aliveMessage.ID}] = updateMember(aliveMessage.MemberMessage, Alive)
+		return true, nil
+	}
+	// Check incarnation
+	if aliveMessage.Incarnation < existingMem.Incarnation {
+		return false, ErrIncarnation
+	}
```
Comment (junbeomlee) [inline member_map.go:198]:
```
1. {Alive Ml, inc=i} overrides
     - {Suspect Ml, inc=j}, i>j
     - {Alive Ml, inc=j}, i>j
```

According to the override rule above, if i == j, we do not override it, so we need to keep the existing status.

File: member_map_internal_test.go [hunk 4]
```diff
@@ -22,6 +22,130 @@ import (
...
+			ID:          "1",
+			Incarnation: 1,
+		},
+	})
+
+	//Then
+	assert.Nil(t, err)
+	assert.Equal(t, true, isChanged)
+	assert.Equal(t, Alive, m.members[MemberID{ID: "1"}].Status)
+
```
Comment (junbeomlee) [inline member_map_internal_test.go:48]: `assert.Equal(t, Alive, m.members[MemberID{ID: "1"}].Status)` this should be
`assert.Equal(t, Suspected, m.members[MemberID{ID: "1"}].Status)`

File: member_map.go [hunk 5]
```diff
@@ -108,26 +121,7 @@ func NewMemberMap() *MemberMap {
...
-	// Remove first k membersID.
-	kMembers := m.waitingMembers[:k]
-	m.waitingMembers = m.waitingMembers[k:len(m.waitingMembers)]
-
-	if len(m.waitingMembers) == 0 {
-		m.waitingMembers = resetWaitingMembersID(m.members)
-	}
-
-	return kMembers
+	return nil
```
Comment (hihiboss) [inline member_map.go:123]: Why erase the implementation and leave just `return nil`?

Comment (junbeomlee) [reply member_map.go:123]: Because we decide to implement this function in another way.

File: member_map.go [hunk 6]
```diff
@@ -90,15 +103,15 @@ func (m *Member) GetID() MemberID {
 
 type MemberMap struct {
 	lock    sync.RWMutex
-	members map[MemberID]Member
+	members map[MemberID]*Member
```
Comment (zeroFruit) [inline member_map.go:105]: oh! good feeling about changing to pointer :+1:

File: member_map.go [hunk 7]
```diff
@@ -198,36 +170,56 @@ func override(newMem Member, existingMem Member) Member {
...
-		waitingMembersID = append(waitingMembersID, member)
+	// if Member is not exist in MemberList
+	if !ok {
+		m.members[MemberID{aliveMessage.ID}] = createMember(aliveMessage.MemberMessage, Alive)
+		return true, nil
 	}
+	// Check incarnation
+	if aliveMessage.Incarnation <= existingMem.Incarnation {
+		return false, ErrIncarnation
+	}
```
Comment (zeroFruit) [inline member_map.go:198]: In my opinion, receiving `aliveMessage.Incarnation <= existingMem.Incarnation` is not an error. it's just old message. how about just logging?

**Comment by junbeomlee** [comment]
LGTM

**Review by junbeomlee — APPROVED** [review]

**Pull request merged**